tests/benchmarks/test_qiskit_adv.py

```diff
@@ -0,0 +1,219 @@
+)
+
+
+################################################################################
+# Pauli String Apply
+################################################################################
+qiskit_params_sm = [
+    pytest.param(nq, m) for nq, m in it.product([2, 4, 6, 8, 10], ["qiskit"])
+]
+qiskit_params_big = [
+    pytest.param(nq, m, marks=extra_benchmark_mark)
+    for nq, m in it.product([12, 14], ["qiskit"])
+]
+
+fast_pauli_params_sm = [
+    pytest.param(nq, m) for nq, m in it.product([2, 4, 6, 8, 10], ["fast_pauli"])
+]
+
+fast_pauli_params_big = [
+    pytest.param(nq, m, marks=extra_benchmark_mark)
+    for nq, m in it.product([12, 14, 16, 18, 20, 22, 24], ["fast_pauli"])
+]
+
+
+@pytest.mark.parametrize(
+    "n_qubits, method",
+    qiskit_params_sm + qiskit_params_big + fast_pauli_params_sm + fast_pauli_params_big,
+)
+def test_pauli_string_apply(benchmark: Callable, n_qubits: int, method: str) -> None:
+    """Benchmark applying a Pauli String to a state."""
+    np.random.seed(18)
+
+    all_pauli_strings = [
+        "".join(s) for s in it.combinations_with_replacement("IXYZ", n_qubits)
+    ]
+    pauli_string = np.random.choice(all_pauli_strings, size=1)[0]
+    print(f"n_qubits: {n_qubits}, method: {method}, pauli_string: {pauli_string}")
+
+    if method == "qiskit":
+        v = Statevector(np.random.random(2**n_qubits).astype(np.complex128))
+        ps = Pauli(pauli_string)
+
+        def f() -> None:
+            v.evolve(ps)
+
+    else:
+        v = np.random.random(2**n_qubits).astype(np.complex128)
+        ps = fp.PauliString(pauli_string)
+
+        def f() -> None:
+            ps.apply(v)
+
+    benchmark(f)
+
+
+################################################################################
+# Pauli Op Apply
+################################################################################
+qiskit_params_sm = [
+    pytest.param(nq, ns, m)
+    for nq, ns, m in it.product([2, 8, 10], [10, 100, 1000], ["qiskit"])
+]
+qiskit_params_big = [
+    pytest.param(nq, ns, m, marks=extra_benchmark_mark)
+    for nq, ns, m in it.product([12, 13, 14, 15], [10, 100, 1000], ["qiskit"])
+]
+
+fast_pauli_params_sm = [
+    pytest.param(nq, ns, m)
+    for nq, ns, m in it.product([2, 8, 10], [10, 100, 1000], ["fast_pauli"])
+]
+
+fast_pauli_params_big = [
+    pytest.param(nq, ns, m, marks=extra_benchmark_mark)
+    for nq, ns, m in it.product(
+        [12, 13, 14, 15, 16, 18], [10, 100, 1000], ["fast_pauli"]
+    )
+]
+
+
+@pytest.mark.parametrize(
+    "n_qubits, n_pauli_strings, method",
+    qiskit_params_sm + qiskit_params_big + fast_pauli_params_sm + fast_pauli_params_big,
+)
+def test_pauli_op_apply(
+    benchmark: Callable,
+    pauli_strings_with_size: Callable,
+    n_qubits: int,
+    n_pauli_strings: int,
+    method: str,
+) -> None:
+    """Benchmark applying a Pauli Operator (in the general sense) to a state."""
+    np.random.seed(18)
+
+    all_pauli_strings = [
```

Comment:
I may be very wrong about this, but I thought Pauli Strings are not necessarily commutative? Not sure if i misunderstand what `it.combinations_with_replacement` does though.
If what i said was right then this could do it
 ```
all_pauli_strings = [
    "".join(s) for s in it.product("IXYZ", repeat=n_qubits)]
```

Reply:
I don't think it matters for the benchmark, but you're absolutely right that there are more unique pauli strings than the ones I'm generating here. I will change this over